From 4cf8d9ef9efcd28853a254cbb846f00c49a5405f Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 30 Nov 2023 15:09:51 +0100
Subject: [PATCH 0091/1695] plat-stm32mp1: set voltage list at pmic driver init

Bound stm32mp1_pmic supported voltage levels list to min/max voltage
level values set from the DT. This change free quite a bit of byte in
the heap for this platform.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git e1bf377b695a729d17ed101c1e254f44b6a13a32 plat-stm32mp1: set voltage list at pmic driver init]
Change-Id: I4a2437bd883bd7eb96fa8a05242a53455123c418
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from a7990eb02b82e88f8474d7704713a6c207962cfb]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../arm/plat-stm32mp1/drivers/stm32mp1_pmic.c | 22 ++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
index c256dd917..42735b898 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
@@ -397,7 +397,8 @@ static int cmp_int_value(const void *a, const void *b)
 	return CMP_TRILEAN(*ia, *ib);
 }
 
-static size_t refine_levels_array(size_t count, int *levels_uv)
+static size_t refine_levels_array(size_t count, int *levels_uv,
+				  int min_uv, int max_uv)
 {
 	size_t n = 0;
 	size_t m = 0;
@@ -413,8 +414,21 @@ static size_t refine_levels_array(size_t count, int *levels_uv)
 			m++;
 		}
 	}
+	count = m + 1;
 
-	return m + 1;
+	for (n = count; n; n--)
+		if (levels_uv[n - 1] <= max_uv)
+			break;
+	count = n;
+
+	for (n = 0; n < count; n++)
+		if (levels_uv[n] >= min_uv)
+			break;
+	count -= n;
+
+	memmove(levels_uv, levels_uv + n, count * sizeof(*levels_uv));
+
+	return count;
 }
 
 static TEE_Result pmic_list_voltages(struct regulator *regulator,
@@ -443,7 +457,9 @@ static TEE_Result pmic_list_voltages(struct regulator *regulator,
 		for (n = 0; n < level_count; n++)
 			levels[n] = level_ref[n] * 1000;
 
-		level_count = refine_levels_array(level_count, levels);
+		level_count = refine_levels_array(level_count, levels,
+						  regulator->min_uv,
+						  regulator->max_uv);
 
 		/* Shrink levels array to not waste heap memory */
 		levels2 = realloc(levels, sizeof(*levels) * level_count);
-- 
2.47.2

